### DISCUSSION TITLE
Java to C/C++ exchange data objects

### DISCUSSION THREAD

**Discussion by mswiderski** [oracle/graal#3634] [Q&A] [answered]
Hello,

I am trying to make it work to call C/C++ from java by passing data. I am using GraalVM 21.2.0 and using the docker image with llvm toolchain installed.

Here is my simple C program

````
#include <stdio.h>

struct Point {
  int x;
  int y;
};

int printHello(struct Point point) {
  FILE *f = fopen("file.txt", "w");
  if (f == NULL) {
      printf("Error opening file!\n");
  } else {
      const char *text = "world";
      fprintf(f, "Hello: %s\n", text);
      fclose(f);
  }
return point.x * point.y;
}
`````

This file is compiled using ` clang -O1 -c -emit-llvm Sample.c`

Next, here is the Java code that uses it

````
import org.graalvm.polyglot.*;
import org.graalvm.polyglot.proxy.*;
import java.io.*;
import java.util.*;

class SampleJava {
public static class Point {
            public int x;
            public int y;

            private Point(int x, int y) {
                this.x = x;
                this.y = y;
            }
        }

    public static void main(String[] args) throws Exception {
        Context polyglot = Context.newBuilder().
                                       allowAllAccess(true).build();
        File file = new File("Sample");
        Source source = Source.newBuilder("llvm", file).build();
        Value cpart = polyglot.eval(source);

        //Map<String, Object> data = new HashMap<>();
        //data.put("x", 10);
        //data.put("y", 20);
        //ProxyObject map = ProxyObject.fromMap(data);

        //System.out.println("Returned from c : " + cpart.getMember("printHello").execute(map));
        System.out.println("Returned from c : " + cpart.getMember("printHello").execute(new Point(4,2)));
    }
}
````

I tried two approached to pass data from Java to C
- using class that match the members of the C struct
- using `ProxyObject.fromMap`

Unfortunately both end up with the same exception

````
Exception in thread "main" Cannot convert com.oracle.truffle.llvm.runtime.interop.LLVMTypedForeignObject@1adb7478 to native pointer.
        at <llvm> printHello(Another.c:8:53-88)
        at org.graalvm.sdk/org.graalvm.polyglot.Value.execute(Value.java:830)
        at SampleJava.main(Sample.java:30)
````

Any ideas how to resolve this? Or other hints how to pass data from Java to C (and back) that are of more complex types?

Thanks

**Comment by rschatz** [comment, marked as answer]
The problem is that foreign objects are always interpreted as pointers in C (that's the closest thing to a Java "reference" we have). You can't pass a Java object to a function that expects a by-value struct. The runtime would have to copy it to replicate by-value semantics. That is currently not implemented (and it probably also wouldn't be what you want, the resulting copy wouldn't be a Java object anymore).

Modifying the `printHello` function to take a pointer as argument should fix it.

**Reply by mswiderski** [reply]
indeed, this solved the problem and now it works as expected, much appreciated @rschatz 

here is a complete c source for completeness

````
#include <stdio.h>

struct Point {
  int x;
  int y;
};

int printHello(struct Point *point) {
  FILE *f = fopen("file.txt", "w");
  if (f == NULL) {
      printf("Error opening file!\n");
  } else {
      const char *text = "world";
      fprintf(f, "Hello: %s\n", text);
      fclose(f);
  }
  return point -> x * point -> y;
}
````

**Reply by mswiderski** [reply]
It seems I can't mark it as answered but @rschatz is certainly correct answer to this question.

**Comment by mswiderski** [comment]
@rschatz one more question related to this - how should I pass string values from Java to C program?

Extending the program in following way

````
#include <stdio.h>

struct Point {
  int x;
  int y;
  char *data;
};

int printHello(struct Point *point) {
  FILE *f = fopen("file.txt", "w");
  if (f == NULL) {
      printf("Error opening file!\n");
  } else {
      const char *text = "world";
      fprintf(f, "Hello: %s, %s\n", text, point -> data);
      fclose(f);
  }
  return point -> x * point -> y;
}
````

and java part

````
import org.graalvm.polyglot.*;
import org.graalvm.polyglot.proxy.*;
import java.io.*;
import java.util.*;

class SampleJava {
     public static class Point {
            public int x;
            public int y;
            public String data;

            private Point(int x, int y, String data) {
                this.x = x;
                this.y = y;
                this.data = data;
            }
        }

    public static void main(String[] args) throws Exception {
        Context polyglot = Context.newBuilder().
                                       allowAllAccess(true).build();
        File file = new File("sample.o");
        Source source = Source.newBuilder("llvm", file).build();
        Value cpart = polyglot.eval(source);

        System.out.println("Returned from c : " + cpart.getMember("printHello").execute(new Point(4,2,"java")));
    }
}
````

fails with following exception

````
Exception in thread "main" Cannot convert java to native pointer.
        at <llvm> printHello(Unknown)
        at org.graalvm.sdk/org.graalvm.polyglot.Value.execute(Value.java:830)
        at SampleJava.main(SampleJava.java:32)
````

**Reply by rschatz** [reply]
Unfortunately we do not support implicit conversion to string. The problem is that there is no way to know whether a `char *` is a pointer to a string, or just a pointer to a byte array (remember, `char` in C is also used as the generic byte-sized integer). It also doesn't help that in LLVM bitcode, the representation of `char*` and `void*` looks exactly the same.

You have to tell the runtime explicitly what you want here. `point->data` returns a generic (untyped) polyglot value. Then you can use the builtin `polyglot_as_string` to convert it to a C string, given an encoding. See https://github.com/oracle/graal/blob/ec2efa31f1236578c89428d3fcbcef48795b5819/sulong/projects/com.oracle.truffle.llvm.libraries.graalvm.llvm/include/graalvm/llvm/polyglot.h#L496.
